### PR TITLE
Block Viewer: Add stats for blocks and last refreshed

### PR DESCRIPTION
**Pull request by kunal-kushwaha**
Signed-off-by: Kunal Kushwaha <kunalkushwaha453@gmail.com>

fix (partially) #3112

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

ss

## Changes
Added stats: 
- Last Refresh
- Total Number of Blocks 
- Total Number of Blocks in Each Group (Hover)

### CODE REVIEW

File: .bingo/go.mod
Comment (bwplotka) [inline .bingo/go.mod:3]: ?

Comment (bwplotka) [reply .bingo/go.mod:3]: Do we need this change?

Also this file was accidently created, we need to remove it ):

**Review by bwplotka — APPROVED** [review]
From user perspective it's LGTM (: 

One comment only. Waiting for @prmsrswt approval though!

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (prmsrswt) [inline pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:69]: Minor nit. This `div` seems redundant, as we essentially have
```html
<div>
  <div>
    {........}
  </div>
</div>
```
So the outer div is just creating an unnecessary div soup.

**Review by prmsrswt — APPROVED** [review]
Just a minor nit along with what Bartek mentioned. LGTM!

**Comment by prmsrswt** [comment]
Also, it'll be nice if you can add some tests :)

File: pkg/ui/react-app/package.json
Comment (prmsrswt) [inline pkg/ui/react-app/package.json:49]: Using a high level UI library just for tooltips is a bit too much. It will [increase the bundle size by a lot](https://bundlephobia.com/result?p=semantic-ui-react@2.0.1) too.

We are already using Reactstrap all over. Use https://reactstrap.github.io/components/tooltips/

File: pkg/ui/react-app/src/thanos/pages/blocks/SourceView.tsx
Comment (prmsrswt) [inline pkg/ui/react-app/src/thanos/pages/blocks/SourceView.tsx:5]: Use https://reactstrap.github.io/components/tooltips/ as we already use Reactstrap all over.

**Comment by prmsrswt** [comment]
This PR contains a lot of unrelated commits, maybe something went iffy during rebase? Anyways the outcome looks solid and I'd definitely want to merge it. @kunal-kushwaha can you clean up the PR so that we can get this merged? Let me know if you need any help with this.

**Comment by kunal-kushwaha** [comment]
> This PR contains a lot of unrelated commits, maybe something went iffy during rebase? Anyways the outcome looks solid and I'd definitely want to merge it. @kunal-kushwaha can you clean up the PR so that we can get this merged? Let me know if you need any help with this.

Oops, will send an update soon.

**Comment by kunal-kushwaha** [comment]
@prmsrswt Just need to update the tests now. Been getting this error: 
```
Summary of all failing tests
 FAIL  src/thanos/pages/blocks/Blocks.test.tsx
  ● Blocks › when data is returned › renders sources

    expect(received).toHaveLength(expected)

    Expected length: 8
    Received length: 0
    Received object: {}

      29 |
      30 |       const sourceViews = blocks.find(SourceView);
    > 31 |       expect(sourceViews).toHaveLength(8);
         |                           ^
      32 |     });
      33 |
      34 |     it('fetched data with different view', async () => {

      at Object.<anonymous> (src/thanos/pages/blocks/Blocks.test.tsx:31:27)

  ● Blocks › when data is returned › fetched data with different view

    expect(received).toHaveLength(expected)

    Expected length: 8
    Received length: 0
    Received object: {}

      40 |
      41 |       const sourceViews = blocks.find(SourceView);
    > 42 |       expect(sourceViews).toHaveLength(8);
         |                           ^
      43 |     });
      44 |   });
      45 |

      at Object.<anonymous> (src/thanos/pages/blocks/Blocks.test.tsx:42:27)

 FAIL  src/thanos/pages/blocks/SourceView.test.tsx
  ● Blocks SourceView › encountered a declaration exception

    The target 'countToolTip' could not be identified in the dom, tip: check spelling

      20 |   };
      21 |
    > 22 |   const sourceView = mount(<SourceView {...defaultProps} />);
         |                      ^
      23 |
      24 |   it('renders a paragraph with title', () => {
      25 |     const title = sourceView.find('div > Tooltip').first();
```

**Comment by kunal-kushwaha** [comment]
@prmsrswt all greeeeeen

**Comment by kunal-kushwaha** [comment]
@prmsrswt @bwplotka Ready for review 😄
giphy

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (prmsrswt) [inline pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:82]: We actually don't need `ready`, `setReady` and the corresponding `useEffect`, using `linkRef.current` directly here would suffice.

```suggestion
            {linkRef.current && (
```

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (prmsrswt) [inline pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:24]:
```suggestion
  const linkRef = useRef<HTMLSpanElement>();
```

Let's try using the proper types if possible. :)

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (prmsrswt) [inline pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:84]: Let's show the full date in the tooltip, not just the time.

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (prmsrswt) [inline pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:27]:
```suggestion
  const toggle = () => setOpen((o) => !o);
```

The state setter functions are async in nature, using it like a closure (accessing the state from outside) can result in stale updates/ race conditions. Take a look at https://reactjs.org/docs/hooks-reference.html#functional-updates

**Review by prmsrswt — COMMENTED** [review]
Looks solid, just a few suggestions.

**Comment by kunal-kushwaha** [comment]
@onprem requested changes made

**Review by bwplotka — DISMISSED** [review]
Looks amazing! LGTM from my side. (:

**Comment by bwplotka** [comment]
Waiting for last approve from @onprem

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (onprem) [inline pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:1]:
```suggestion
import React, { FC, useMemo, useRef, useState } from 'react';
```

Sorry if I missed it but I guess we are not using `useEffect` anywhere so let's remove the import.

File: pkg/ui/react-app/src/thanos/pages/blocks/SourceView.tsx
Comment (onprem) [inline pkg/ui/react-app/src/thanos/pages/blocks/SourceView.tsx:46]:
```suggestion
          {linkRef.current && (
```

same as the last review, we can simply skip the `ready` state and the corresponding `useEffect` by directly using `linkRef.current` here.

**Review by onprem — COMMENTED** [review]
I think you missed the changes in the second file. I should have made an explicit comment there as well, sorry for that.

Let's get these small changes done and then we'll merge the PR. :)

File: pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx
Comment (kunal-kushwaha) [reply pkg/ui/react-app/src/thanos/pages/blocks/Blocks.tsx:1]: Oops I missed that, thanks!

**Comment by kunal-kushwaha** [comment]
It automatically "dismissed stale reviews" when I pushed new commits 👀

**Comment by onprem** [comment]
Oh looks like we missed updating `bindata.go`, Can you run `make assets` and then push the changes?

**Comment by kunal-kushwaha** [comment]
Hey @onprem 
Been getting this error, tried resolving it by changing dependency versions but no luck. 

```
$ react-scripts build
Creating an optimized production build...
Failed to compile.

./src/utils/index.ts
  Line 0:  Parsing error: Cannot read property 'map' of undefined
```

Can you maybe check if `make assets` is working for you with the PR when you get time?

**Comment by kunal-kushwaha** [comment]
@onprem requested changes made, all greeen ✅

**Comment by kunal-kushwaha** [comment]
@onprem Let me know when we can merge this I'll then resolve the merge conflicts 👍

**Comment by onprem** [comment]
Hey, sorry of all the delay :sweat: 

Can you rebase it now? Let's get this in as we are planning to do a release soon.

One interesting thing to keep in mind during rebase is that now we have Dark Mode in the UI, so you might want to verify your styles stay consistent with the theme.

**Comment by kunal-kushwaha** [comment]
@onprem resolved merge conflicts

**Comment by yeya24** [comment]
@onprem Is there still interest to get this in?

**Pull request closed without merging**
